src/witness/witnessUtil.ml

```diff
@@ -79,7 +79,61 @@ struct
       emit_after_lock
     else
       emit_other
+
+  let find_syntactic_loop_head n =
+    let prevs = Cfg.prev n in
+    List.find_map (fun (edges, prev) ->
+        let stmts = Cfg.skippedByEdge prev edges n in
+        List.find_map (fun s ->
+            match s.GoblintCil.skind with
+            | Loop (_, loc, _, _, _) -> Some loc
+            | _ -> None
+          ) stmts
+      ) prevs
+end
+
+module YamlInvariant (FileCfg: MyCFG.FileCfg) =
+struct
+  include Invariant (FileCfg)
+
+  let is_stub_node n =
+    let fundec = Node.find_fundec n in
+    Cil.hasAttribute "goblint_stub" fundec.svar.vattr
+
+  let location_location (n : Node.t) = (* great naming... *)
+    match n with
+    | Statement s ->
+      let {loc; _}: CilLocation.locs = CilLocation.get_stmtLoc s in
+      if not loc.synthetic && is_invariant_node n && not (is_stub_node n) then (* TODO: remove is_invariant_node? *)
+        Some loc
+      else
+        None
+    | FunctionEntry _ | Function _ ->
+      (* avoid FunctionEntry/Function, because their locations are not inside the function where asserts could be inserted *)
+      None
+
+  let is_invariant_node n = Option.is_some (location_location n)
+
+  let loop_location n =
+    find_syntactic_loop_head n
+    |> BatOption.filter (fun _loc -> not (is_stub_node n))
+
+  let is_loop_head_node n = Option.is_some (loop_location n)
+end
+
+module YamlInvariantValidate (FileCfg: MyCFG.FileCfg) =
+struct
+  include Invariant (FileCfg)
+
+  (* TODO: filter synthetic?
+
+     Almost all loops are transformed by CIL, so the loop constructs all get synthetic locations. Filtering them from the locator could give some odd behavior: if the location is right before the loop and all the synthetic loop head stuff is filtered, then the first non-synthetic node is already inside the loop, not outside where the location actually was.
+     Similarly, if synthetic locations are then filtered, witness.invariant.loop-head becomes essentially useless.
+     I guess at some point during testing and benchmarking I achieved better results with the filtering removed. *)
+
+  let is_loop_head_node = NH.mem loop_heads
 end
+[@@deprecated]
```

Comment:
Where is this still used?

Reply:
It's not anymore, but I don't want to immediately remove it because there's a long comment explaining some old behavior with possible practical considerations. Now validation uses the same predicates as generation, which seems the reasonable thing to do, but without tests for YAML witness validation yet, I'm not completely sure whether this should be removed or there's some truth to this and the distinction needs to be reintroduced.

This PR has been driven by making Goblint's generated witnesses easier to consume by Ultimate, but does not consider the effect on Goblint Validator validating witnesses from all the other tools. There could be regressions in that regard so I don't want to completely forget this point.